pkg/pod/entrypoint_lookup.go

```diff
@@ -43,70 +51,57 @@ type EntrypointCache interface {
 // Images that are not specified by digest will be specified by digest after
 // lookup in the resulting list of containers.
 func resolveEntrypoints(ctx context.Context, cache EntrypointCache, namespace, serviceAccountName string, steps []corev1.Container) ([]corev1.Container, error) {
-	// Keep a local cache of name->image lookups, just for the scope of
+	// Keep a local cache of name->imageData lookups, just for the scope of
 	// resolving this set of steps. If the image is pushed to before the
-	// next run, we need to resolve its digest and entrypoint again, but we
+	// next run, we need to resolve its digest and commands again, but we
 	// can skip lookups while resolving the same TaskRun.
-	localCache := map[name.Reference]v1.Image{}
+	localCache := map[name.Reference]imageData{}
 	for i, s := range steps {
-		if len(s.Command) != 0 {
-			// Nothing to resolve.
+
+		// If the command is already specified, there's nothing to resolve.
+		if len(s.Command) > 0 {
 			continue
 		}
 
-		origRef, err := name.ParseReference(s.Image, name.WeakValidation)
+		ref, err := name.ParseReference(s.Image, name.WeakValidation)
 		if err != nil {
 			return nil, err
 		}
-		var img v1.Image
-		if cimg, found := localCache[origRef]; found {
-			img = cimg
+		var id imageData
+		if cid, found := localCache[ref]; found {
+			id = cid
 		} else {
-			// Look it up in the cache. If it's not found in the
-			// cache, it will be resolved from the registry.
-			img, err = cache.Get(ctx, origRef, namespace, serviceAccountName)
+			// Look it up for real.
+			lid, err := cache.Get(ctx, ref, namespace, serviceAccountName)
 			if err != nil {
 				return nil, err
 			}
-			// Cache it locally in case another step specifies the same image.
-			localCache[origRef] = img
-		}
+			id = *lid
 
-		ep, digest, err := imageData(origRef, img)
-		if err != nil {
-			return nil, err
+			// Cache it locally in case another step in this task specifies the same image.
+			localCache[ref] = *lid
 		}
 
-		cache.Set(digest, img) // Cache the lookup for next time this image is looked up by digest.
+		// Resolve the original reference to a reference by digest.
+		steps[i].Image = ref.Context().Digest(id.digest.String()).String()
 
-		steps[i].Image = digest.String() // Specify image by digest, since we know it now.
-		steps[i].Command = ep            // Specify the command explicitly.
+		if len(id.commands) == 1 {
+			// Promote the single found command to the step's command.
+			for _, v := range id.commands {
+				steps[i].Command = v
+				break
+			}
+		} else {
```

Comment:
Generally my bias is towards having a single way of passing this information.  This special case is probably 80+% of cases, so we're really limiting our coverage of the else path.  My personal bias would be to send everything through the `else` path and shake out any bugs more aggressively.

Reply:
Yeah, you're right. This exposes another tiny opportunity for a bug due to missing information in the image config, but it's tiny, and having this exercised better would be good.